### DISCUSSION TITLE
Custom gitlens.remotes does not show up in selection

### DISCUSSION THREAD

**Discussion by simbabque** [gitkraken/vscode-gitlens#1366] [Q&A]
I've set up a custom `gitlens.remotes` object for a local install of Phabricator, which I want to use to browse commits and share links with coworkers. This is based on the documentation the dialog sends me to at https://github.com/eamodio/vscode-gitlens#remote-provider-integration-settings-.

```json
    "gitlens.integrations.enabled": true,
    "gitlens.remotes": [{
        "domain": "phabricator.example.org",
        "type": "Custom",
        "name": "Phabricator",
        "protocol": "https",
        "urls": {
            "repository":   "https://phabricator.example.org/diffusion/Foo/",
            "commit":       "https://phabricator.example.org/rFoo${id}",
            "branch":       "https://phabricator.example.org/diffusion/Foo/browse/${branch}/",
            "branches":     "https://phabricator.example.org/diffusion/Foo/branches/",
            "file":         "https://phabricator.example.org/diffusion/Foo/browse/master/${file}",
            "fileInBranch": "https://phabricator.example.org/diffusion/Foo/browse/${branch}/${file}",
            "fileInCommit": "https://phabricator.example.org/diffusion/Foo/browse/${branch}/${file};${id}",
            "fileLine":     "$${line}",
            "fileRange":    "$${start}-${end}"
        }
    }]
```
However when I look at the file history and click the _Open File on Remote_ button (:globe_with_meridians:), the dialog tells me _No auto-detected or configured remote providers found_.

I've restarted VSCode, and I've fiddled with different settings, but I can't get it to show up. If I do this with a public GitHub repository (where I've allowed gitlens to connect to GH) it works without any further settings and the dialog shows all the git remotes I've set up in that repository.

I can't find any debug output, so I don't know what else I can try.

**Comment by eamodio** [comment]
Please enable output channel logging by setting `"gitlens.outputLevel": "debug"` in your settings.

This will enable logging to the *GitLens* channel in the *Output* pane. Once enabled, please reproduce the issue, and attach the log from the *GitLens* channel.

Can you also provide me with what your remote looks like in you `.git/config` file?

**Reply by simbabque** [reply]
Ah, I should have looked for a log level in the settings. Thanks. I've turned on debugging, but it doesn't output anything related to remotes when I click the _Open File on Remote_. All I see is the `git blame` on the file when the file history pane opens. 

(Note I've changed all URLs and file names to examples but I made sure to keep them different from each other if they're not equal in reality).

```
[2021-02-09 15:00:30:937] [/home/simbabque/code] git status --porcelain=v2 --find-renames -- foo/bar.html • 74 ms 
```

However I think I might have misunderstood how this feature works. I don't have this website set up as a remote in my `.git/config`. We don't use it as the _source of truth_, so to speak, like you would with github. We only use it as a repository browser and for code reviews. It operates on a copy of the real remote repository that gets updated via a receive hook on the real remote, which is outside my control. I can't push code to this thing, for example.

This is the relevant part of my `.git/config`:

```
[remote "origin"]
        fetch = +refs/heads/*:refs/remotes/origin/*
        url = git://foo-git.example.org/git/foo.git
```


What I wanted to use this particular feature in GitLens for was just be able to get a link to a file or line I am looking at in my IDE easily so I can post that into chat or tickets to share with my team. I suspect that that's not the main aim of this feature.

I've now tried to set the `name` property of the remote in the GitLens config to `origin` (was `Phabricator` above), but that doesn't make it show up either. I do however see debug output after saving that change in the config file and switching to a different code file.

```
[2021-02-09 15:09:09:451] [/home/simbabque/code] git ls-files -- foo/qrr.js • 23 ms 
[2021-02-09 15:09:09:473] [/home/simbabque/code] git ls-files -- foo/qrr.js • 20 ms 
[2021-02-09 15:09:09:484] [/home/simbabque/code] git remote -v • 11 ms 
[2021-02-09 15:09:09:802] [/home/simbabque/code] git blame --root --incremental -- foo/qrr.js • 125 ms 
[2021-02-09 15:09:09:815] [/home/simbabque/code] git for-each-ref --format=%3ch%3e%(HEAD)%3cn%3e%(refname)%3cu%3e%(upstream:short)%3ct%3e%(upstream:track)%3cr%3e%(objectname)%3cd%3e%(committerdate:iso8601) refs/heads refs/remotes • 125 ms 
[2021-02-09 15:09:09:892] [/home/simbabque/code] git status --porcelain=v2 --find-renames -- foo/qrr.js • 58 ms 
[2021-02-09 15:09:10:334] [/home/simbabque/code] git log --format=%x3c%x2ff%x3e%n%x3cr%x3e%x20%H%n%x3ca%x3e%x20%aN%n%x3ce%x3e%x20%aE%n%x3cd%x3e%x20%at%n%x3cc%x3e%x20%ct%n%x3cp%x3e%x20%P%n%x3cs%x3e%n%B%n%x3c%x2fs%x3e%n%x3cf%x3e -n41 -m --numstat --summary -- foo/qrr.js • 491 ms 
```

**Reply by eamodio** [reply]
Ah, all those remote features in GitLens are based on matching a built-in or custom configuration to a Git remote. So in order for those features to work, you'd have to add that server/service as a Git remote. You could probably add it as a `fetch` only remote (even though I know you don't want to fetch from it either).

**Reply by simbabque** [reply]
Brilliant, that worked!

I've changed the `name` back to `phabricator` and created a fetch-only remote with this command (`Foo` is the project name).

```
$ git remote add --mirror=fetch phabricator https://phabricator.example.org/diffusion/Foo/
```

It now wants to open file history URLs, but there's one more issue. It uses `fileInCommit` for the file history, which has `{branch}` in it.

```
            "fileInCommit": "https://phabricator.example.org/diffusion/Foo/browse/${branch}/${file};${id}",
```

The `{branch}` comes out empty, so the URL I get is wrong.

```
                                                    V
https://phabricator.example.org/diffusion/Foo/browse//lib/Bar.pm;d34db33f
```
I can pin this to `master`, which is probably fine as we don't have old branches kicking around, but I was wondering if that's something you might want to know about. 

I am currently on `master` while doing this. Here's the log output, after which the "do you wan't to open this URL" window pops upen.

```
[2021-02-10 11:10:34:502] [/home/simbabque/code] git status --porcelain=v2 --find-renames -- lib/Bar.pm • 59 ms 
[2021-02-10 11:10:34:781] [/home/simbabque/code] git blame --root --incremental -- lib/Bar.pm • 354 ms 
[2021-02-10 11:10:34:945] [/home/simbabque/code] git log --format=%x3c%x2ff%x3e%n%x3cr%x3e%x20%H%n%x3ca%x3e%x20%aN%n%x3ce%x3e%x20%aE%n%x3cd%x3e%x20%at%n%x3cc%x3e%x20%ct%n%x3cp%x3e%x20%P%n%x3cs%x3e%n%B%n%x3c%x2fs%x3e%n%x3cf%x3e -n41 -m --numstat --summary -- lib/Bar.pm • 490 ms 
```
I've run all three commands manually. The first one doesn't output anything. The second one looks like a nice machine-readable log that does not contain the word `master`. The third one has the actual commits, looks correct, but no branch info either.

Thanks a lot for your help and this awesome plugin. I'm only just discovering what it can do, and I've learned a lot about git debugging this as well. :)

**Reply by rdnlsmith** [reply]
I've been struggling with this configuration as well. This discussion has gotten me _almost_ there.

My first problem was I misunderstood `domain` and `protocol`.

* Domain: I had entered `{server}.{local domain}.local` but I had cloned the repo with just `{server}`, and these needed to match exactly.
* Protocol: I thought it was "protocol for the URLs" (https) when it was actually "protocol you cloned via" (ssh). In retrospect that seems like it should have been obvious to me.

Fixing those got it to recognize that I had a remote configured and send the appropriate URLs to the browser. However, I now have a second problem.

This is an on-premises TFS instance, so the URLs all have the format

```
https://{server}:{port}/tfs/DefaultCollection/{project}/_git/...
```

I was surprised to discover that `${repo}` produces `tfs/DefaultCollection/{project}/_git/{repo name}`, and further `${repoBase}` produces `tfs` while `${repoPath}` is the rest of it. This wouldn't be an issue except, to my everlasting dismay, the project name has a space in it. When I cloned the repo, the space was encoded as `%20` in my git settings, and now when I generate a file URL as

```
https://{server}:{port}/${repo}?path=${file}`
```

it re-encodes the `%20` so it becomes `%2520`, which obviously doesn't work.

Is this the same problem as #796?

Are there any other repo-related tokens, or any other way I could extract just the name of the repo (last segment of `${repoPath}`)? In which case I could specify

```
https://{server}:{port}/tfs/DefaultCollection/{project}/_git/${repoName}?path=${file}
```

Thanks!

**Comment by Anutrix** [comment]
Facing the same issue.
No auto-detected or configured remote providers found.

I want to use Gitlens with Mozilla Phabricator: https://phabricator.services.mozilla.com/source/mozilla-central/ and tried:
```
"gitlens.integrations.enabled": true,
"gitlens.remotes": [
    {
        "domain": "https://phabricator.services.mozilla.com",
        "type": "Custom",
        "name": "Mozilla Phabricator",
        "urls": {
            "repository": "https://phabricator.services.mozilla.com/source/mozilla-central/",
            "commit": "https://phabricator.services.mozilla.com/rmozilla-central${id}",
            "branch": "https://phabricator.services.mozilla.com/source/mozilla-central/browse/${branch}/",
            "branches": "https://phabricator.services.mozilla.com/source/mozilla-central/branches/",
            "file": "https://phabricator.services.mozilla.com/source/mozilla-central/browse/master/${file}",
            "fileInBranch": "https://phabricator.services.mozilla.com/source/mozilla-central/browse/${branch}/${file}",
            "fileInCommit": "https://phabricator.services.mozilla.com/source/mozilla-central/browse/${branch}/${file};${id}",
            "fileLine": "$${line}",
            "fileRange": "$${start}-${end}"
        }
    }
],
```

Doesn't seem to work and debug logs just shows:
```
025-02-27 15:20:43.680 [info] [2025-02-27 09:50:43.168] [GIT   *4084ms] [k:/mozilla-source/mozilla-unified] git for-each-ref --format=%3ch%3e%(HEAD)%3cn%3e%(refname)%3cu%3e%(upstream:short)%3ct%3e%(upstream:track)%3cr%3e%(objectname)%3cd%3e%(committerdate:iso8601) refs/heads refs/remotes [*4084ms] (slow)
```
This is just used to show list of branches.


Replacing domain line with `"regex": "hg::https://hg.mozilla.org/mozilla-unified",` and then trying to open remote file gives:
```
2025-02-27 15:47:30.758 [info] [2025-02-27 10:17:30.255] OpenOnRemoteCommand
TypeError: Cannot read properties of undefined (reading 'indexOf')
```
